_drafts/tagless-akka.md

```diff
@@ -0,0 +1,88 @@
+---
+layout: post
+title: "Tagless final and akka-http"
+categories: tips
+---
+
+Have you ever wondered how could you set up some akka-http routes for your tagless final program? Did you end up hardcoding your concrete `F[_]` to something `ToResponseMarshallable` in order to build your routes? I have. I did. Not anymore!
+
+# The problem
+
+It turns out I have already written bits about tagless final [here] and [there], I just didn't know it's called that then.
+I think we can dig up that little example from these posts.
+
+```scala
+trait Database[F[_]] {
+  def load(id: Int): F[User]
+  def save(user: User): F[Unit]
+}
+```
+
+Now assume we want to provide an HTTP endpoint for loading users by id. Also assume that we can effortlessly convert case classes to JSON so we don't have to deal with it now. Something like this:
+
+```scala
+def route[F[_] : Database] = get {
```

Comment:
I would use the alternative akka http directive syntax (`&` operator) as imho it promotes a better style, i.e.:
```
get & path("users" / IntNumber) { id =>
  complete(Database[F].load(id))
}
```

Reply:
I don't like it that much, because you have to parenthesise them

```scala
(get & path("users" / IntNumber)) { id =>
  complete(Database[F].load(id))
}
```

If it worked without parentheses, I'd be in :)